jetbrains-core/src/software/aws/toolkits/jetbrains/core/stack/View.kt

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.core.stack
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.Disposable
+import com.intellij.ui.AnimatedIcon
+import com.intellij.util.Alarm
+import com.intellij.util.AlarmFactory
+import software.amazon.awssdk.services.cloudformation.model.ResourceStatus
+import software.amazon.awssdk.services.cloudformation.model.StackStatus
+import javax.swing.Icon
+import javax.swing.SwingUtilities
+
+enum class StatusType(val icon: Icon, private val animatedIconStrategy: (() -> AnimatedIcon)? = null) {
+    UNKNOWN(AllIcons.Actions.Clear),
+    PROGRESS(AllIcons.Process.ProgressResume, { AnimatedIcon.Grey() }),
+    COMPLETED(AllIcons.Process.State.GreenOK),
+    DELETED(AllIcons.Process.State.YellowStr),
+    FAILED(AllIcons.Process.State.RedExcl);
+
+    val animatedIconIfPossible: Icon get() = animatedIconStrategy?.let { it() } ?: icon
+}
+
+internal val StackStatus.type: StatusType
+    get() = when (this) {
+        StackStatus.UPDATE_COMPLETE, StackStatus.CREATE_COMPLETE -> StatusType.COMPLETED
+        StackStatus.UPDATE_IN_PROGRESS,
+        StackStatus.CREATE_IN_PROGRESS,
+        StackStatus.ROLLBACK_IN_PROGRESS,
+        StackStatus.UPDATE_ROLLBACK_IN_PROGRESS,
+        StackStatus.DELETE_IN_PROGRESS,
+        StackStatus.REVIEW_IN_PROGRESS,
+        StackStatus.UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS,
+        StackStatus.UPDATE_COMPLETE_CLEANUP_IN_PROGRESS -> StatusType.PROGRESS
+        StackStatus.DELETE_COMPLETE -> StatusType.DELETED
+        StackStatus.ROLLBACK_COMPLETE,
+        StackStatus.DELETE_FAILED,
+        StackStatus.UPDATE_ROLLBACK_FAILED,
+        StackStatus.ROLLBACK_FAILED,
+        StackStatus.UPDATE_ROLLBACK_COMPLETE,
+        StackStatus.CREATE_FAILED -> StatusType.FAILED
+        else -> StatusType.UNKNOWN
+    }
+
+internal val ResourceStatus.type: StatusType
+    get() = when (this) {
+        ResourceStatus.CREATE_COMPLETE, ResourceStatus.UPDATE_COMPLETE -> StatusType.COMPLETED
+        ResourceStatus.CREATE_FAILED, ResourceStatus.UPDATE_FAILED, ResourceStatus.DELETE_FAILED -> StatusType.FAILED
+        ResourceStatus.DELETE_COMPLETE, ResourceStatus.DELETE_SKIPPED -> StatusType.DELETED
+        ResourceStatus.CREATE_IN_PROGRESS, ResourceStatus.DELETE_IN_PROGRESS, ResourceStatus.UPDATE_IN_PROGRESS -> StatusType.PROGRESS
+        else -> StatusType.UNKNOWN
+    }
+
+/**
+ * @property update Function0<Unit> function to call to update icon
+ */
+internal data class IconInfo(val icon: Icon, val update: () -> Unit)
+
+internal interface ViewWithIcons {
+    fun getIconsAndUpdaters(): Collection<IconInfo>
+}
+
+/**
+ * Animates icon
+ * @property views Array<out ViewWithIcons> views with animated icons
+ */
+internal class IconAnimator(private val interval: Int, private vararg val views: ViewWithIcons) : Disposable {
```

Comment:
Kind of surprised to see this isn't a util in AnimatecIcon

Reply:
Unfortunatelly there is no API for that in platform. I asked our platform guys.